src/Symfony/Component/Cache/Adapter/AbstractAdapter.php

```diff
@@ -286,44 +275,50 @@ public function saveDeferred(CacheItemInterface $item)
      */
     public function commit()
     {
-        $f = $this->mergeByLifetime;
-        $ko = array();
+        $ok = true;
+        $byLifetime = $this->mergeByLifetime;
+        $byLifetime = $byLifetime($this->deferred, $this->namespace);
+        $ko = $this->deferred = array();
 
-        foreach ($f($this->deferred, $this->namespace) as $lifetime => $values) {
+        foreach ($byLifetime as $lifetime => $values) {
             try {
-                if (true === $ok = $this->doSave($values, $lifetime)) {
-                    continue;
-                }
+                $e = $this->doSave($values, $lifetime);
             } catch (\Exception $e) {
-                $ok = false;
             }
-            if (false === $ok) {
-                $ok = array_keys($values);
+            if (true === $e || array() === $e) {
+                continue;
             }
-            foreach ($ok as $id) {
-                $ko[$lifetime][] = array($id => $values[$id]);
+            if (is_array($e) || 1 === count($values)) {
```

Comment:
don't retry when only one item or when the adapter returned an explicit failing subset